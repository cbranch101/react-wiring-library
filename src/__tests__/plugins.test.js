import React, {Fragment, useState} from 'react'
import {queryHelpers} from '@testing-library/react'
import {getRender} from '../index'
const {queryAllByAttribute} = queryHelpers
const Icon = ({name, onClick}) => <span onClick={onClick} xlinkHref={name} />

const Root = () => {
  return (
    <div data-custom-value="foo" data-testid="root">
      Root
    </div>
  )
}

const fixture = <Root />

const wiring = {
  children: {
    root: {
      findValue: 'root',
      serialize: () => 'Root',
    },
  },
}

const functionReduceTest = () => {
  const plugins = [
    {
      getFuncs: funcs => {
        return {two: () => funcs.one()}
      },
    },
    {
      getFuncs: funcs => {
        return {two: () => 'foo', three: () => funcs.two()}
      },
    },
  ]

  const getGetAllFunctions = () => {
    return plugins.reduceRight(
      (memo, plugin) => {
        const {getFuncs} = plugin
        return funcs => {
          const updatedFuncs = {
            ...funcs,
            ...getFuncs(funcs),
          }
          return memo(updatedFuncs)
        }
      },
      funcs => funcs,
    )
  }
  const getAllFunctions = getGetAllFunctions()
  const {one, two, three} = getAllFunctions({one: () => 'test'})
  console.log({
    one: one(),
    two: two(),
    three: three(),
  })
}

functionReduceTest()

describe('Plugins', () => {
  describe('when getWithinElementFunctions is added to a plugin', () => {
    test.skip('the returned functions should be added to all find nodes', async () => {
      const customGlobal = () => {}
      const config = {
        plugins: [
          {
            getWithinElementFunctions: element => {
              return {
                getCustomValue: () => element.getAttribute('data-custom-value'),
              }
            },
          },
        ],
      }
      const render = getRender(wiring, config)
      const {findRoot, customGlobalFunction} = render(fixture)
      expect(customGlobalFunction).toEqual(customGlobal)
      const {root, getCustomValue} = await findRoot()
      console.log(getCustomValue)
      expect(root).toMatchSnapshot('initial render')
    })
  })
})
