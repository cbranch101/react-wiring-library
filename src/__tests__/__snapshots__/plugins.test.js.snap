// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Queries custom queries should work with all built in functions: initial render 1`] = `Root`;

exports[`Plugins when getCustomFunctions is added to a plugin returned global functions should be returned from render: initial render 1`] = `Root`;

exports[`Plugins when getCustomFunctions is added to a plugin: initial render 1`] = `Root`;
